miri-script/src/commands.rs

```diff
@@ -0,0 +1,711 @@
+        Ok(Self { libdir, sysroot })
+    }
+}
+
+impl<'a> MiriRunner<'a> {
+    pub(super) fn run(command: &'a super::Subcommands) -> Result<()> {
+        let miri_dir = detect_miri_dir()?;
+        let active_toolchain = get_current_toolchain(&miri_dir)?;
+        let config = command.get_auto_config(&miri_dir);
+        let cargo_extra_flags = std::env::var("CARGO_EXTRA_FLAGS").unwrap_or_default();
+        let extra_flags = BTreeMap::new();
+
+        let mut ret = Self {
+            miri_dir,
+            active_toolchain,
+            command,
+            extra_flags,
+            extra_vars: Default::default(),
+            cargo_extra_flags,
+        };
+        if let Some(config) = config {
+            ret.extra_flags.insert("MIRI_AUTO_OPS".into(), "42".into());
+            if config.toolchain {
+                ret.toolchain(&[])?;
+            }
+            if config.fmt {
+                ret.fmt(&[])?;
+            }
+            if config.clippy {
+                ret.clippy(&["--".to_owned(), "-D".to_owned(), "warnings".to_owned()])?;
+            }
+            if config.toolchain {
+                // Let's make sure to actually use that toolchain, too.
+                // We do it after running all the auto things though to use a detected toolchain.
+                ret.active_toolchain = "miri".to_owned();
+            }
+        }
+        let extra_vars = Extras::new(&ret.active_toolchain)?;
+        if !extra_vars.libdir.exists() {
+            println!("Something went wrong determining the library dir.");
+            println!("I got {} but that does not exist.", extra_vars.libdir.display());
+            println!("Please report a bug at https://github.com/rust-lang/miri/issues.");
+            std::process::exit(2);
+        }
+        ret.extra_vars = extra_vars;
+        // Share target dir between `miri` and `cargo-miri`.
+        let target_dir = std::env::var("CARGO_TARGET_DIR")
+            .ok()
+            .filter(|val| !val.is_empty())
+            .unwrap_or_else(|| {
+                let target_dir = path!(ret.miri_dir / "target");
+                target_dir.to_string_lossy().to_string()
+            });
+        ret.extra_flags.insert("CARGO_TARGET_DIR".to_owned(), target_dir);
+
+        // We configure dev builds to not be unusably slow.
+        let devel_opt_level = std::env::var("CARGO_PROFILE_DEV_OPT_LEVEL")
+            .ok()
+            .filter(|val| !val.is_empty())
+            .unwrap_or_else(|| "2".to_owned());
+        ret.extra_flags.insert("CARGO_PROFILE_DEV_OPT_LEVEL".to_owned(), devel_opt_level);
+
+        let rustflags = std::env::var("RUSTFLAGS").unwrap_or_default();
+        let rustflags = format!(
+            "-Zunstable-options -Wrustc::internal -Wrust_2018_idioms -Wunused_lifetimes -Wsemicolon_in_expressions_from_macros {rustflags}"
+        );
+        let rustflags =
+            format!("-C link-args=-Wl,-rpath,{} {rustflags}", ret.extra_vars.libdir.display());
```

Comment:
Please port over the comments from the shell script as well (also applies anywhere else if you dropped any comments)